File: Java/trappingRainWater.java
class Solution {
    public int trap(int[] height) {
        int[][] talls = new int[height.length][2];

        int tallestLeft = 0;
        for(int i=0; i<height.length; i++){
            talls[i][0] = tallestLeft;
            tallestLeft = Math.max(tallestLeft, height[i]);
        }

        int tallestRight = 0;
        for(int i=height.length-1; i>=0; i--){
            talls[i][1] = tallestRight;
            tallestRight = Math.max(tallestRight, height[i]);
        }

        int totalWater = 0;
        for(int i=0; i<talls.length; i++){
            int currentWater = Math.min(talls[i][0], talls[i][1])-height[i];
            totalWater += currentWater>0? currentWater : 0;
        }

        return totalWater;
    }
}

File: Python/trappingRainwater.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) <= 2:
            return 0
        maxLeft = height[0]
        maxRight = height[-1]

        left = 0
        right = len(height)-1
        water = 0
        
        while (left < right):
            maxLeft = max(maxLeft, height[left])
            maxRight = max(maxRight, height[right])

            if maxLeft < maxRight:
                water += max(0, maxLeft-height[left])
                left += 1
            
            else:
                water += max(0, maxRight-height[right])
                right -= 1

        return water